Sources/InteroperabilityTests/InteroperabilityTestCases.swift

```diff
@@ -0,0 +1,721 @@
+          case .trailingMetadata(let receivedTrailingMetadata):
+            try self.checkTrailingMetadata(receivedTrailingMetadata)
+          default:
+            try assertFailure(
+              "The client should have received only a message and trailing metadata."
+            )
+          }
+        }
+      case .failure(_):
+        try assertFailure("The client should have received a response from the server.")
+      }
+    }
+  }
+}
+
+/// This test verifies unary calls succeed in sending messages, and propagate back status code and
+/// message sent along with the messages.
+///
+/// Server features:
+/// - UnaryCall
+/// - FullDuplexCall
+/// - Echo Status
+///
+/// Procedure:
+/// 1. Client calls UnaryCall with:
+///    ```
+///    {
+///        response_status:{
+///            code: 2
+///            message: "test status message"
+///        }
+///    }
+///    ```
+/// 2. Client calls FullDuplexCall with:
+///    ```
+///    {
+///        response_status:{
+///            code: 2
+///            message: "test status message"
+///        }
+///    }
+///    ```
+/// 3. and then half-closes
+///
+/// Client asserts:
+/// - received status code is the same as the sent code for both Procedure steps 1 and 2
+/// - received status message is the same as the sent message for both Procedure steps 1 and 2
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct StatusCodeAndMessage: InteroperabilityTest {
+  let expectedCode = 2
+  let expectedMessage = "test status message"
+
+  func run(client: GRPCClient) async throws {
+    let testServiceClient = Grpc_Testing_TestService.Client(client: client)
+
+    let message = Grpc_Testing_SimpleRequest.with {
+      $0.responseStatus = Grpc_Testing_EchoStatus.with {
+        $0.code = Int32(self.expectedCode)
+        $0.message = self.expectedMessage
+      }
+    }
+
+    try await testServiceClient.unaryCall(request: ClientRequest.Single(message: message)) {
+      response in
+      switch response.accepted {
+      case .failure(let error):
+        try assertEqual(error.code.rawValue, self.expectedCode)
+        try assertEqual(error.message, self.expectedMessage)
+      case .success(_):
+        try assertFailure(
+          "The client should receive an error with the status code and message sent by the client."
+        )
+      }
+    }
+
+    let request = ClientRequest.Stream { writer in
+      let message = Grpc_Testing_StreamingOutputCallRequest()
+      try await writer.write(message)
+    }
+
+    try await testServiceClient.fullDuplexCall(request: request) { response in
+      switch response.accepted {
+      case .failure(let error):
+        try assertEqual(error.code.rawValue, self.expectedCode)
+        try assertEqual(error.message, self.expectedMessage)
+      case .success(_):
+        try assertFailure(
+          "The client should receive an error with the status code and message sent by the client."
+        )
+      }
```

Comment:
There are two places the server can fail a streaming RPC: before processing it and during processing it. Because this test relies on the service consuming the request to know it should fail the RPC we know that the RPC will be accepted. We therefore need to consume the messages in the response and catch the error there:

```swift
do {
for try await message in response.messages {
  throw AssertionFailure(...)
} catch let error as RPCError {
  // validate error code and message
} 

```